pkg/controller/pxc/users_custom.go

```diff
@@ -0,0 +1,307 @@
+package pxc
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+
+	api "github.com/percona/percona-xtradb-cluster-operator/pkg/apis/pxc/v1"
+	"github.com/percona/percona-xtradb-cluster-operator/pkg/pxc/users"
+)
+
+func (r *ReconcilePerconaXtraDBCluster) reconcileCustomUsers(ctx context.Context, cr *api.PerconaXtraDBCluster) error {
+	if cr.Spec.Users == nil && len(cr.Spec.Users) == 0 {
+		return nil
+	}
+
+	if cr.Status.Status != api.AppStateReady {
+		return nil
+	}
+
+	log := logf.FromContext(ctx)
+
+	internalSecrets := corev1.Secret{}
+	err := r.client.Get(context.TODO(),
+		types.NamespacedName{
+			Namespace: cr.Namespace,
+			Name:      internalSecretsPrefix + cr.Name,
+		},
+		&internalSecrets,
+	)
+	if err != nil && !k8serrors.IsNotFound(err) {
+		return errors.Wrap(err, "get internal sys users secret")
+	}
+
+	um, err := getUserManager(cr, &internalSecrets)
+	if err != nil {
+		return err
+	}
+	defer um.Close()
+
+	sysUserNames := sysUserNames()
+
+	for _, user := range cr.Spec.Users {
+		if _, ok := sysUserNames[user.Name]; ok {
+			log.Error(nil, "creating user with reserved user name is forbidden", "user", user.Name)
+			continue
+		}
+
```

Comment:
should on this stage check the user name for mysql limitations for names. 
- 32 symbols
- Special Characters
- reserved words 
The first two can be checked by regexp.

Reply:
> if I delete the user from CR should it be deleted from DB?

No. Pls check the spec listed in the ticket to check all the cases.

Reply:
> should we on this stage check the user name for mysql limitations for names.
> 
> * 32 symbols
> * Special Characters
> * reserved words
>   The first two can be checked by regexp.

I don't think so, didn't do that for PSMDB as well.